### PR TITLE
core/tracker: add tracker component

### PR DESCRIPTION
**Pull request by dB2510**
Adds tracker component with deadliner interface. Tests would be implemented in subsequent PRs alongwith Deadliner implementation.

category: feature
ticket: #768

### CODE REVIEW

File: core/tracker/tracker.go [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+		scheduler:         "scheduler",
+		fetcher:           "fetcher",
+		consensus:         "consensus",
+		validatorAPI:      "validatorAPI",
+		parSigDBInternal:  "parSigDBInternal",
+		parSigEx:          "parSigEx",
+		parSigDBThreshold: "parSigDBThreshold",
+		sigAgg:            "sigAgg",
+	}[c]
+}
```
Comment (corverroos) [inline core/tracker/tracker.go:54]: can replace with `//go generate stringer -type=component`

File: core/tracker/tracker.go [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+			t.analyzeDuty(ctx, duty)
+			t.trimDuty(duty)
+		}
+	}
+}
+
+// storeEvent stores the event as value with the duty as key. Also, adds duty to the deadliner.
+func (t *Tracker) storeEvent(e event) {
+	t.events[e.duty] = append(t.events[e.duty], e)
+	t.deadliner.Add(e.duty)
```
Comment (corverroos) [inline core/tracker/tracker.go:107]: can al oinline in `Run`, since this doesn't do much

File: core/tracker/tracker.go [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+	events := t.events[duty]
+
+	// Sort in reverse order (see order above).
+	sort.Slice(events, func(i, j int) bool {
+		return events[i].component > events[j].component
+	})
+
+	// Case of failed duties
+	if events[0].component != sigAgg {
```
Comment (corverroos) [inline core/tracker/tracker.go:123]: avoid nesting: 
```
if events[0].component == sigAgg {
  // Duty completed succesfully
  return
}
```

File: core/tracker/tracker.go [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+		log.Error(ctx, "Duty stuck", errors.New("duty stuck"), z.Str("component", failedComponent), z.Str("duty", duty.String()))
+
+		return
+	}
+	// TODO(dhruv): Case of cluster participation (duty success)
+}
+
+// trimDuty trims the events for a given duty.
+func (t *Tracker) trimDuty(duty core.Duty) {
+	delete(t.events, duty)
```
Comment (corverroos) [inline core/tracker/tracker.go:135]: can inline since this doesn't do much

File: core/tracker/tracker.go [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+		default:
+			t.input <- event{
+				duty:      duty,
+				component: scheduler,
+				pubkey:    pubkey,
+			}
+		}
+	}
+
+	log.Debug(ctx, "Sent events to tracker", z.Str("component", scheduler.string()))
```
Comment (corverroos) [inline core/tracker/tracker.go:153]: this will spam logs, suggest removing all these

File: core/tracker/tracker.go [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+// storeEvent stores the event as value with the duty as key. Also, adds duty to the deadliner.
+func (t *Tracker) storeEvent(e event) {
+	t.events[e.duty] = append(t.events[e.duty], e)
+	t.deadliner.Add(e.duty)
+}
+
+// analyzeDuty analyzes the events for a given duty after the duty's deadline is exceeded.
+func (t *Tracker) analyzeDuty(ctx context.Context, duty core.Duty) {
+	// Deadline crossed for this duty.
+	log.Debug(ctx, "Deadline exceeded", z.Str("duty", duty.String()))
```
Comment (corverroos) [inline core/tracker/tracker.go:113]: Suggest only logging failures

File: core/tracker/tracker.go [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+}
+
+// storeEvent stores the event as value with the duty as key. Also, adds duty to the deadliner.
+func (t *Tracker) storeEvent(e event) {
+	t.events[e.duty] = append(t.events[e.duty], e)
+	t.deadliner.Add(e.duty)
+}
+
+// analyzeDuty analyzes the events for a given duty after the duty's deadline is exceeded.
+func (t *Tracker) analyzeDuty(ctx context.Context, duty core.Duty) {
```
[the remark this answers is not in the document]
Comment (corverroos) [reply core/tracker/tracker.go:111]: Then report via a testable function call:
```
failed, failedComponent, failedMsg := analyseFailedDauty(duty, t.events[duty])
t.reportFailedDuty(duty, failed, failedComponent, failedMsg)
```

File: core/validatorapi/validatorapi.go [hunk 8]
```diff
@@ -217,6 +218,11 @@ func (c *Component) RegisterAwaitBeaconBlock(fn func(ctx context.Context, slot i
 	c.awaitBlockFunc = fn
 }
 
+// RegisterTracker registers the tracker component.
+func (c *Component) RegisterTracker(fn func(context.Context, core.Duty, core.ParSignedDataSet)) {
```
Comment (corverroos) [inline core/validatorapi/validatorapi.go:222]: no need for this, can use existing `RegisterParSigDB` (should probably rename to `Subscribe`)

**Review by corverroos — APPROVED** [review]

File: core/interfaces.go [hunk 9]
```diff
@@ -207,7 +207,7 @@ func Wire(sched Scheduler,
 		VAPIRegisterAwaitAttestation:    vapi.RegisterAwaitAttestation,
 		VAPIRegisterGetDutyDefinition:   vapi.RegisterGetDutyDefinition,
 		VAPIRegisterPubKeyByAttestation: vapi.RegisterPubKeyByAttestation,
-		VAPIRegisterParSigDB:            vapi.RegisterParSigDB,
+		VAPIRegisterParSigDB:            vapi.Subscribe,
```
Comment (corverroos) [inline core/interfaces.go:210]: should also rename this `VAPIRegisterParSigDB`

File: core/tracker/tracker.go [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case e := <-t.input:
+			// TODO(dhruv): do not store events for expired duties
+			t.events[e.duty] = append(t.events[e.duty], e)
+			t.deadliner.Add(e.duty)
+		case duty := <-t.deadliner.C():
+			failed, failedComponent, failedMsg := analyseFailedDuty(duty, t.events[duty])
+			if failed {
```
Comment (corverroos) [inline core/tracker/tracker.go:87]: you should probably call report in all cases, so you can test positive and negative cases. The report function can just ignore non-failed.

File: core/tracker/tracker.go [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+			}
+			// TODO(dhruv): Case of cluster participation (duty success)
+			// t.analyseClusterParticipation()
+			delete(t.events, duty)
+		}
+	}
+}
+
+// analyseFailedDuty analyzes the events for a given duty after the duty's deadline is exceeded.
+func analyseFailedDuty(duty core.Duty, events []event) (bool, component, string) {
```
Comment (corverroos) [inline core/tracker/tracker.go:98]: suggest improve godoc to indicate what it returns, which isn't clear from the function or comment

File: core/tracker/tracker.go [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+	// Sort in reverse order (see order above).
+	sort.Slice(events, func(i, j int) bool {
```
Comment (corverroos) [inline core/tracker/tracker.go:100]: note this modifies the original events map which isn't great, suggest coping the slice before modifying it.

File: core/tracker/tracker.go [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+		return events[i].component > events[j].component
+	})
+
+	if events[0].component == sigAgg {
```
Comment (corverroos) [inline core/tracker/tracker.go:104]: this will panic if the slice is empty

File: core/tracker/tracker.go [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+		// Duty completed successfully
+		return false, sigAgg, ""
+	}
+
+	return true, events[0].component + 1, fmt.Sprintf("%s is failed in %s component", duty.String(), (events[0].component + 1).String())
```
Comment (corverroos) [inline core/tracker/tracker.go:109]: I think we can improve the error message. Be more specific. If it got stuck during validatorAPI, we can say "the VC didn't submit a signed duty", if it got stuck at parsigdb then "peers didn't share any/sufficient partial signatures", etc.

File: core/tracker/tracker.go [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+
+func (*Tracker) reportFailedDuties(core.Duty, component, string) {
+	// TODO(dhruv): instrument failed duty
+}
+
+// SchedulerEvent inputs event from core.Scheduler component.
+func (t *Tracker) SchedulerEvent(_ context.Context, duty core.Duty, defSet core.DutyDefinitionSet) error {
+	for pubkey := range defSet {
+		select {
+		case <-t.quit:
```
Comment (corverroos) [inline core/tracker/tracker.go:120]: should also add case for ctx timeout

File: core/tracker/tracker.go (hunk 12 above)
Comment (dB2510) [reply core/tracker/tracker.go:100]: yeah right, slices are pass by reference

File: core/tracker/tracker.go (hunk 14 above)
Comment (xenowits) [reply core/tracker/tracker.go:109]: Agreed.

We can have specific analysis in a separate PR.

**Pull request merged**